magento/magento2: Fixes for the schema cache.xsd

andrewbess opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

This PR fixes the schema "cache.xsd" from Magento\Framework\Cache.

1. The attribute "instance" has changed on required because if a developer won't be to use it new cache type will not be created;
2. The parameters "label" and "description" have changed on required because of the class "Magento\Framework\App\Cache\TypeList" gets theirs without checking for existing. Please take a look at [this code](https://github.com/magento/magento2/blob/cde4f67d8e02ad7ffe1de5051010bdce71fdabfd/lib/internal/Magento/Framework/App/Cache/TypeList.php#L124);
3. Checking on unique of the type name has added;
4. The configuration reader class has fixed for more informative during to run CLI commands;
5. The class SchemaLocator has been fixed to check the cache configuration during configuration merging;
6. The unnecessary checking has removed from the class TypeList;
7. Unit tests have been covered.

### Fixed Issues (if relevant)

Fixes: https://github.com/magento/magento2/issues/26224

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

andrewbess committed:
magento/magento2: Fixes for the schema cache.xsd

swnsma reviewed (changes requested):
Hi @andrewbess!
Thank you for your contribution!
Could you please add unit tests to check new schema validation?
Maybe, we also should refactor the class \Magento\Framework\App\Cache\TypeList as 'instance' will become a required parameter? We may simplify some checks here.

ihor-sviziev reviewed (changes requested):
Please see following review details: https://github.com/magento/magento2/pull/26933#pullrequestreview-362147607

andrewbess committed:
magento/magento2: Additional fixes for the cache configuration

andrewbess committed:
magento/magento2: Additional fixes for the cache configuration

andrewbess commented:
> Hi @andrewbess!
> Thank you for your contribution!
> Could you please add unit tests to check new schema validation?
> Maybe, we also should refactor the class \Magento\Framework\App\Cache\TypeList as 'instance' will become a required parameter? We may simplify some checks here.

Hello @swnsma 
Thank you for review.
I added the changes.
There are additional changes for schema, tests, additional changes for TypeList in the PR.
Also, I have changed description of PR.
Please recheck.
Thank you in advance.

rogyar commented on lib/internal/Magento/Framework/Cache/Config/Reader.php:
```diff
@@ -45,4 +50,25 @@ public function __construct(
+    {
+        try {
+            return parent::read($scope);
+        } catch (\Magento\Framework\Exception\LocalizedException $e) {
```
May I ask you why do we need to transform this localized exception and throw once again?

Thank you

andrewbess replied:
Hello @rogyar.
Thank you for your remark.
Previously, I have added the render-method that prepares an error message to output error messages correctly.
But it caused other problems.
I have removed new methods.
Please recheck.
Thank you in advance.

swnsma commented on lib/internal/Magento/Framework/Cache/Test/Unit/XsdTest.php:
```diff
@@ -0,0 +1,65 @@
+ */
+namespace Magento\Framework\Cache\Test\Unit;
+
+class XsdTest extends \PHPUnit\Framework\TestCase
```
Hi @andrewbess,

I have discovered that we already have bunch of tests which test *.xsd files.
They are located in Magento\Test\Integrity namespace.
Could I ask you to move your your tests under this namespace to keep consistency in tests structure?
Please check \Magento\Test\Integrity\Magento\Widget\WidgetConfigTest for example.

andrewbess replied:
Hello @swnsma.
I have moved tests by the suggested path.
Also, I added declaration strict types where it is required.
Please recheck.
Thank you in advance.

swnsma reviewed (changes requested):
Please check my comment. 
Would you be so kind to check strict types declaration, type hints and add 'declare(strict_types=1);' to changed files.
Thank you!

swnsma commented on dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php:
```diff
@@ -0,0 +1,86 @@
+     * @dataProvider schemaCorrectlyIdentifiesInvalidXmlDataProvider
+     */
+    public function testSchemaCorrectlyIdentifiesInvalidXml(
+        $xmlString,
```
Please add strict type 'string'.

swnsma commented on dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * @var \Magento\Framework\TestFramework\Unit\Utility\XsdValidator
+     */
+    protected $_xsdValidator;
```
Please
**protected -> private**
**$_xsdValidator -> $xsdValidator**

swnsma commented on dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * @var \Magento\Framework\Config\Dom\UrnResolver
+     */
+    protected $urnResolver;
```
Please
**protected -> private**

swnsma commented on dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php:
```diff
@@ -0,0 +1,86 @@
+     * Path to xsd schema file
+     * @var string
+     */
+    protected $_xsdSchema;
```
Please
**protected -> private**
**$_xsdSchema -> $xsdSchema**

swnsma commented on lib/internal/Magento/Framework/Cache/Exception/WrongConfigurationException.php:
```diff
@@ -0,0 +1,16 @@
+ *
+ * @api
+ */
+class WrongConfigurationException extends \Exception
```
Sorry, I can not find place where this kind of exception is used...
Could you advice?

swnsma commented on lib/internal/Magento/Framework/App/Test/Unit/Cache/TypeListTest/Type.php:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Cache type "Type"
+ */
+class Type extends TagScope
```
I am not sure if we need class to only handle constants.
Can we just use private constants in places where they need to be used?

swnsma reviewed (changes requested):
Hi @andrewbess!
Thank you for your contribution!
Could you please check my notices?

swnsma committed:
Merge branch '2.4-develop' into improvement/cache-schema

swnsma committed:
Merge branch '2.4-develop' into improvement/cache-schema

Commit:
Merge branch 'improvement/cache-schema' of https://github.com/andrewbess/magento2 into improvement/cache-schema

 Conflicts:
	lib/internal/Magento/Framework/App/Test/Unit/Cache/TypeListTest.php

andrewbess commented:
Closed in favor of https://github.com/magento/magento2/pull/27307
